sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,207 @@
+                                        int num_pads,
+                                        ptrdiff_t base_register_offset,
+                                        int attr_fields_per_reg,
+                                        bitfield_field32_t *field_out,
+                                        ptrdiff_t *register_offset_out) {
+  if (index >= num_pads) {
+    return false;
+  }
+
+  uint32_t register_index;
+
+  register_index = index / attr_fields_per_reg;
+
+  field_out->index =
+      (index % PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG) * PADCTRL_PARAM_ATTRDW;
+
+  field_out->mask = kPadctrlAttrFieldMask;
+
+  *register_offset_out =
+      base_register_offset + (kPadctrlRegisterWidthInBytes * register_index);
+
+  return true;
+}
+
+/**
+ * Helper function for dif_padctrl_enable_attr, reads back an attribute bit to
+ * see if was set, assumes all arguments are valid.
+ */
+static dif_padctrl_enable_attr_result_t test_attr_set_success(
+    const dif_padctrl_t *padctrl, ptrdiff_t pad_register_offset,
+    uint32_t attr_bit_index) {
+  bool test_attr = mmio_region_get_bit32(padctrl->base_addr,
+                                         pad_register_offset, attr_bit_index);
+
+  // Attribute fields are WARL so if attribute isn't supported read value
+  // won't have it set.
+  if (test_attr) {
+    return kDifPadctrlEnableOk;
+  } else {
+    return kDifPadctrlEnableAttrNotSupported;
+  }
+}
+
+dif_padctrl_enable_attr_result_t dif_padctrl_enable_attr(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t attr, bool enable) {
+  if (padctrl == NULL) {
+    return kDifPadctrlEnableBadArg;
+  }
+
+  if (attr > kDifPadctrlAttrLast) {
+    return kDifPadctrlEnableBadArg;
+  }
+
+  bitfield_field32_t attr_field;
+  ptrdiff_t pad_register_offset;
+
+  if (kind == kDifPadctrlPadMio) {
+    if (!calc_pad_attr_field_and_reg(index, PADCTRL_PARAM_NMIOPADS,
+                                     PADCTRL_MIO_PADS0_REG_OFFSET,
+                                     PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG,
+                                     &attr_field, &pad_register_offset)) {
+      return kDifPadctrlEnableBadArg;
+    }
+  } else {
+    if (!calc_pad_attr_field_and_reg(index, PADCTRL_PARAM_NDIOPADS,
+                                     PADCTRL_DIO_PADS0_REG_OFFSET,
+                                     PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG,
+                                     &attr_field, &pad_register_offset)) {
+      return kDifPadctrlEnableBadArg;
+    }
+  }
+
+  // Determine which bit within the register controls the attribute being
+  // modified.
+  uint32_t attr_bit_index = attr_field.index + attr;
+
+  if (enable) {
+    // Get the attribute field from the appropriate register
+    uint32_t attr_reg_value =
+        mmio_region_read32(padctrl->base_addr, pad_register_offset);
+
+    attr_field.value = bitfield_get_field32(attr_reg_value, attr_field);
```

Comment:
Really we need to start migrating bitwise functionality from `mmio` to `bitfield` library, then these manipulations would be entirely done in terms of `bitfield` library.

It can always be done as a follow-up.